File: humanMax-mobile/src/services/notificationService.ts
import { PushNotifications } from '@capacitor/push-notifications';
import { LocalNotifications } from '@capacitor/local-notifications';
import { Capacitor } from '@capacitor/core';
import { Preferences } from '@capacitor/preferences';
import { apiClient } from './apiClient';
import type { Meeting } from '../types';

type NotificationTapCallback = (data: { type: string; meeting_id?: string; [key: string]: any }) => void;

class NotificationService {
  private isInitialized = false;
  private deviceToken: string | null = null;
  private notificationTapCallbacks: NotificationTapCallback[] = [];

  async initialize(): Promise<void> {
    if (this.isInitialized || !Capacitor.isNativePlatform()) {
      return;
    }

    try {
      // Request permission
      let permStatus = await PushNotifications.checkPermissions();

      if (permStatus.receive === 'prompt') {
        permStatus = await PushNotifications.requestPermissions();
      }

      if (permStatus.receive !== 'granted') {
        console.warn('Push notification permission not granted');
        return;
      }

      // Register for push notifications
      await PushNotifications.register();

      // Listen for registration
      PushNotifications.addListener('registration', async (token) => {
        console.log('Push registration success, token: ' + token.value);
        this.deviceToken = token.value;
        await this.saveDeviceToken(token.value);
        await this.registerDeviceWithBackend(token.value);
      });

      // Listen for registration errors
      PushNotifications.addListener('registrationError', (error) => {
        console.error('Error on registration: ' + JSON.stringify(error));
      });

      // Listen for push notifications received
      PushNotifications.addListener('pushNotificationReceived', (notification) => {
        console.log('Push notification received: ', notification);
      });

      // Listen for push notification actions (when user taps notification)
      PushNotifications.addListener('pushNotificationActionPerformed', (action) => {
        console.log('Push notification action performed', action);
        const data = action.notification.data || {};
        this.handleNotificationTap(data);
      });

      this.isInitialized = true;
    } catch (error) {
      console.error('Error initializing notifications:', error);
    }
  }

  private async saveDeviceToken(token: string): Promise<void> {
    await Preferences.set({
      key: 'device_token',
      value: token,
    });
  }

  private async registerDeviceWithBackend(token: string): Promise<void> {
    try {
      // Get user timezone
      const timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
      
      // Get device info
      const platform = Capacitor.getPlatform();
      const deviceInfo = {
        platform,
        appVersion: '1.0.0', // TODO: Get from package.json or Capacitor config
      };

      await apiClient.registerDevice(token, platform, timezone, deviceInfo);
      console.log('Device registered with backend');
    } catch (error) {
      console.error('Error registering device with backend:', error);
      // Don't throw - device registration failure shouldn't break the app
    }
  }

  /**
   * Register a callback for notification taps
   */
  onNotificationTap(callback: NotificationTapCallback): () => void {
    this.notificationTapCallbacks.push(callback);
    // Return unsubscribe function
    return () => {
      const index = this.notificationTapCallbacks.indexOf(callback);
      if (index > -1) {
        this.notificationTapCallbacks.splice(index, 1);
      }
    };
  }

  /**
   * Handle notification tap and notify all callbacks
   */
  private handleNotificationTap(data: any): void {
    console.log('Handling notification tap with data:', data);
    this.notificationTapCallbacks.forEach((callback) => {
      try {
        callback(data);
      } catch (error) {
        console.error('Error in notification tap callback:', error);
      }
    });
  }

  async getDeviceToken(): Promise<string | null> {
    if (this.deviceToken) {
      return this.deviceToken;
    }

    try {
      const { value } = await Preferences.get({ key: 'device_token' });
      return value || null;
    } catch {
      return null;
    }
  }

  async scheduleMeetingReminder(
    meeting: Meeting,
    minutesBefore: number = 15
  ): Promise<void> {
    if (!Capacitor.isNativePlatform()) {
      // For web, use browser notifications
      this.scheduleWebNotification(meeting, minutesBefore);
      return;
    }

    const startTime = meeting.start?.dateTime
      ? new Date(meeting.start.dateTime)
      : meeting.start?.date
      ? new Date(meeting.start.date)
      : null;

    if (!startTime) {
      console.warn('Cannot schedule reminder: no start time');
      return;
    }

    const reminderTime = new Date(startTime.getTime() - minutesBefore * 60 * 1000);
    const now = new Date();

    if (reminderTime <= now) {
      console.warn('Reminder time is in the past');
      return;
    }

    // Schedule local notification
    // Note: Capacitor Push Notifications plugin doesn't support scheduling
    // For scheduling, we'd need to use a native plugin or handle server-side
    // For now, we'll store the reminder info and handle it server-side or via background task
    await this.saveReminder(meeting.id, reminderTime, minutesBefore);
  }

  private async scheduleWebNotification(
    meeting: Meeting,
    minutesBefore: number
  ): Promise<void> {
    if (!('Notification' in window)) {
      return;
    }

    const permission = await Notification.requestPermission();
    if (permission !== 'granted') {
      return;
    }

    const startTime = meeting.start?.dateTime
      ? new Date(meeting.start.dateTime)
      : meeting.start?.date
      ? new Date(meeting.start.date)
      : null;

    if (!startTime) {
      return;
    }

    const reminderTime = new Date(startTime.getTime() - minutesBefore * 60 * 1000);
    const now = new Date();
    const delay = reminderTime.getTime() - now.getTime();

    if (delay <= 0) {
      return;
    }

    setTimeout(() => {
      new Notification(meeting.summary || meeting.title || 'Meeting', {
        body: `Your meeting starts in ${minutesBefore} minutes`,
        icon: '/icon.png',
        tag: `meeting-${meeting.id}`,
      });
    }, delay);
  }

  private async saveReminder(
    meetingId: string,
    reminderTime: Date,
    minutesBefore: number
  ): Promise<void> {
    const reminders = await this.getReminders();
    reminders.push({
      meetingId,
      reminderTime: reminderTime.toISOString(),
      minutesBefore,
    });
    await Preferences.set({
      key: 'meeting_reminders',
      value: JSON.stringify(reminders),
    });
  }

  private async getReminders(): Promise<
    Array<{ meetingId: string; reminderTime: string; minutesBefore: number }>
  > {
    try {
      const { value } = await Preferences.get({ key: 'meeting_reminders' });
      return value ? JSON.parse(value) : [];
    } catch {
      return [];
    }
  }

  async scheduleDailySummary(hour: number = 8, minute: number = 0): Promise<void> {
    // Store preference for daily summary time
    await Preferences.set({
      key: 'daily_summary_time',
      value: JSON.stringify({ hour, minute }),
    });
  }

  async sendNotification(title: string, body: string, data?: any): Promise<void> {
    if (!Capacitor.isNativePlatform()) {
      if ('Notification' in window && Notification.permission === 'granted') {
        new Notification(title, {
          body,
          icon: '/icon.png',
          data,
        });
      }
      return;
    }

    // For native, use Local Notifications
    try {
      const permission = await LocalNotifications.checkPermissions();
      if (permission.display !== 'granted') {
        const request = await LocalNotifications.requestPermissions();
        if (request.display !== 'granted') {
          console.warn('Local notification permission not granted');
          return;
        }
      }

      await LocalNotifications.schedule({
        notifications: [
          {
            title,
            body,
            id: Date.now(),
            schedule: { at: new Date(Date.now() + 100) }, // Show immediately
            sound: 'default',
            attachments: undefined,
            actionTypeId: '',
            extra: data,
          },
        ],
      });
    } catch (error) {
      console.error('Error sending local notification:', error);
    }
  }

  /**
   * Send a test notification (for testing purposes)
   */
  async sendTestNotification(): Promise<void> {
    await this.sendNotification(
      'Test Meeting Reminder',
      'This is a test notification. Your meeting "Team Standup" starts in 15 minutes.',
      { type: 'test', meetingId: 'test-123' }
    );
  }

  /**
   * Schedule a local notification for a meeting reminder
   */
  async scheduleLocalReminder(
    meeting: Meeting,
    minutesBefore: number = 15
  ): Promise<void> {
    if (!Capacitor.isNativePlatform()) {
      this.scheduleWebNotification(meeting, minutesBefore);
      return;
    }

    const startTime = meeting.start?.dateTime
      ? new Date(meeting.start.dateTime)
      : meeting.start?.date
      ? new Date(meeting.start.date)
      : null;

    if (!startTime) {
      console.warn('Cannot schedule reminder: no start time');
      return;
    }

    const reminderTime = new Date(startTime.getTime() - minutesBefore * 60 * 1000);
    const now = new Date();

    if (reminderTime <= now) {
      console.warn('Reminder time is in the past');
      return;
    }

    try {
      const permission = await LocalNotifications.checkPermissions();
      if (permission.display !== 'granted') {
        const request = await LocalNotifications.requestPermissions();
        if (request.display !== 'granted') {
          console.warn('Local notification permission not granted');
          return;
        }
      }

      await LocalNotifications.schedule({
        notifications: [
          {
            title: meeting.summary || meeting.title || 'Meeting Reminder',
            body: `Your meeting starts in ${minutesBefore} minutes`,
            id: parseInt(meeting.id?.replace(/\D/g, '') || '0') || Date.now(),
            schedule: { at: reminderTime },
            sound: 'default',
            attachments: undefined,
            actionTypeId: '',
            extra: {
              type: 'meeting_reminder',
              meetingId: meeting.id,
              minutesBefore,
            },
          },
        ],
      });

      console.log(`Scheduled reminder for meeting "${meeting.summary || meeting.title}" at ${reminderTime.toLocaleString()}`);
    } catch (error) {
      console.error('Error scheduling local notification:', error);
    }
  }
}

export const notificationService = new NotificationService();
export default notificationService;

File: humanMax-mobile/ios/App/App/Services/NotificationService.swift
import Foundation
import UserNotifications
import UIKit

/// Service for managing push notifications and local notifications
@MainActor
class NotificationService: NSObject, ObservableObject {
    static let shared = NotificationService()
    
    @Published var deviceToken: String?
    @Published var isInitialized = false
    
    private let apiClient: APIClient
    private var notificationTapCallbacks: [(NotificationData) -> Void] = []
    
    private override init() {
        self.apiClient = APIClient.shared
        super.init()
    }
    
    // MARK: - Initialization
    
    /// Set up the notification delegate synchronously (must be called early in app lifecycle)
    func setupDelegate() {
        // Set delegate synchronously to ensure notifications are handled even during app startup
        UNUserNotificationCenter.current().delegate = self
    }
    
    /// Initialize notification service (async parts: authorization and registration)
    func initialize() async {
        guard !isInitialized else { return }
        
            // Request authorization
            do {
                let granted = try await requestAuthorization()
                guard granted else {
                    print("Notification permission not granted")
                    return
                }
                
                // Register for remote notifications (only if Push Notifications capability is enabled)
                // This will fail gracefully if Push Notifications is not available (e.g., free Apple Developer account)
                await MainActor.run {
                    // Check if Push Notifications capability is available
                    // If not available, this will fail silently, which is fine for development
                    UIApplication.shared.registerForRemoteNotifications()
                }
                
                isInitialized = true
            } catch {
                print("Error initializing notifications: \(error)")
                // Still mark as initialized if local notifications are available
                // Remote notifications may fail on free developer accounts
                isInitialized = true
            }
    }
    
    /// Request notification authorization
    private func requestAuthorization() async throws -> Bool {
        let center = UNUserNotificationCenter.current()
        let granted = try await center.requestAuthorization(options: [.alert, .sound, .badge])
        return granted
    }
    
    // MARK: - Device Token Management
    
    /// Handle device token registration (called from AppDelegate)
    func didRegisterForRemoteNotifications(deviceToken: Data) {
        let tokenParts = deviceToken.map { data in String(format: "%02.2hhx", data) }
        let token = tokenParts.joined()
        
        self.deviceToken = token
        
        // Save to UserDefaults
        UserDefaults.standard.set(token, forKey: "device_token")
        
        // Register with backend
        Task {
            await registerDeviceWithBackend(token: token)
        }
    }
    
    /// Handle registration error
    func didFailToRegisterForRemoteNotifications(error: Error) {
        print("Failed to register for remote notifications: \(error)")
    }
    
    /// Register device token with backend
    private func registerDeviceWithBackend(token: String) async {
        do {
            let timezone = TimeZone.current.identifier
            let deviceInfo = DeviceInfo(
                platform: "ios",
                appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
            )
            
            _ = try await apiClient.registerDevice(
                deviceToken: token,
                platform: "ios",
                timezone: timezone,
                deviceInfo: deviceInfo
            )
            
            print("Device registered with backend")
        } catch {
            print("Error registering device with backend: \(error)")
        }
    }
    
    /// Get stored device token
    func getDeviceToken() -> String? {
        if let token = deviceToken {
            return token
        }
        return UserDefaults.standard.string(forKey: "device_token")
    }
    
    // MARK: - Local Notifications
    
    /// Send a local notification immediately
    func sendNotification(title: String, body: String, data: [String: Any]? = nil) async throws {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        
        if let data = data {
            content.userInfo = data
        }
        
        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil // nil trigger = immediate
        )
        
        try await UNUserNotificationCenter.current().add(request)
    }
    
    /// Schedule a local notification for a meeting reminder
    func scheduleMeetingReminder(meeting: Meeting, minutesBefore: Int = 15) async throws {
        guard let start = meeting.start,
              let startTime = parseMeetingTime(start) else {
            throw NotificationError.invalidMeetingTime
        }
        
        let reminderTime = startTime.addingTimeInterval(-Double(minutesBefore * 60))
        let now = Date()
        
        guard reminderTime > now else {
            throw NotificationError.reminderTimeInPast
        }
        
        let content = UNMutableNotificationContent()
        content.title = meeting.title ?? meeting.summary
        content.body = "Your meeting starts in \(minutesBefore) minutes"
        content.sound = .default
        content.userInfo = [
            "type": "meeting_reminder",
            "meeting_id": meeting.id,
            "minutes_before": minutesBefore
        ]
        
        let dateComponents = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute],
            from: reminderTime
        )
        
        let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: false)
        
        // Use meeting ID as identifier (convert to numeric if possible)
        let identifier = meeting.id.replacingOccurrences(of: "[^0-9]", with: "", options: .regularExpression)
        let finalIdentifier = identifier.isEmpty ? String(Int(reminderTime.timeIntervalSince1970)) : identifier
        
        let request = UNNotificationRequest(
            identifier: finalIdentifier,
            content: content,
            trigger: trigger
        )
        
        try await UNUserNotificationCenter.current().add(request)
        print("Scheduled reminder for meeting \"\(meeting.title ?? meeting.summary)\" at \(reminderTime)")
    }
    
    /// Schedule daily summary notification
    func scheduleDailySummary(hour: Int = 8, minute: Int = 0) async throws {
        var dateComponents = DateComponents()
        dateComponents.hour = hour
        dateComponents.minute = minute
        
        let content = UNMutableNotificationContent()
        content.title = "Daily Summary"
        content.body = "Your daily meeting summary is ready"
        content.sound = .default
        content.userInfo = [
            "type": "daily_summary"
        ]
        
        let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: true)
        
        let request = UNNotificationRequest(
            identifier: "daily_summary",
            content: content,
            trigger: trigger
        )
        
        try await UNUserNotificationCenter.current().add(request)
        
        // Save preference
        UserDefaults.standard.set(["hour": hour, "minute": minute], forKey: "daily_summary_time")
    }
    
    /// Cancel a scheduled notification
    func cancelNotification(identifier: String) {
        UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [identifier])
    }
    
    /// Cancel all scheduled notifications
    func cancelAllNotifications() {
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
    }
    
    // MARK: - Notification Tap Handling
    
    /// Register callback for notification taps
    func onNotificationTap(_ callback: @escaping (NotificationData) -> Void) -> () -> Void {
        notificationTapCallbacks.append(callback)
        
        // Return unsubscribe function
        return { [weak self] in
            guard let self = self else { return }
            if let index = self.notificationTapCallbacks.firstIndex(where: { $0 as AnyObject === callback as AnyObject }) {
                self.notificationTapCallbacks.remove(at: index)
            }
        }
    }
    
    /// Handle notification tap
    private func handleNotificationTap(data: NotificationData) {
        notificationTapCallbacks.forEach { callback in
            callback(data)
        }
    }
    
    // MARK: - Helpers
    
    /// Parse meeting time from MeetingTime struct
    private func parseMeetingTime(_ meetingTime: MeetingTime?) -> Date? {
        guard let meetingTime = meetingTime else { return nil }
        
        if let dateTimeString = meetingTime.dateTime {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            return formatter.date(from: dateTimeString) ?? ISO8601DateFormatter().date(from: dateTimeString)
        }
        
        if let dateString = meetingTime.date {
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd"
            return formatter.date(from: dateString)
        }
        
        return nil
    }
}

// MARK: - UNUserNotificationCenterDelegate

@MainActor
extension NotificationService: UNUserNotificationCenterDelegate {
    /// Handle notification when app is in foreground
    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void
    ) {
        // Show notification even when app is in foreground
        completionHandler([.banner, .sound, .badge])
    }
    
    /// Handle notification tap
    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {
        let userInfo = response.notification.request.content.userInfo
        
        let data = NotificationData(
            type: userInfo["type"] as? String ?? "unknown",
            meetingId: userInfo["meeting_id"] as? String,
            data: userInfo
        )
        
        handleNotificationTap(data: data)
        completionHandler()
    }
}

// MARK: - Notification Data Model

struct NotificationData {
    let type: String
    let meetingId: String?
    let data: [AnyHashable: Any]
}

// MARK: - Notification Error Types

enum NotificationError: Error, LocalizedError {
    case permissionDenied
    case invalidMeetingTime
    case reminderTimeInPast
    case schedulingFailed
    
    var errorDescription: String? {
        switch self {
        case .permissionDenied:
            return "Notification permission denied"
        case .invalidMeetingTime:
            return "Invalid meeting time"
        case .reminderTimeInPast:
            return "Reminder time is in the past"
        case .schedulingFailed:
            return "Failed to schedule notification"
        }
    }
}
